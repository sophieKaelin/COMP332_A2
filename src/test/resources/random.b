//
// This file is part of COMP332 Assignment 2/3 2019.
//
// weBCPL, a retro BCPL to WebAssembly compiler.
//
// © 2019, Dominic Verity and Anthony Sloane, Macquarie University.
//         All rights reserved.
// © 2005, Martin Richards
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// Random number generator, adapted from example given in the BCPL manual.
// It is based upon an algorithm given in Knuth: The art of programming, 
// vol 2, p 26.
//

MANIFEST { 
    SIZE = 55 
    MASK = #x_FFF_FFFF
    MAGIC_1 = #x_234_5678
    MAGIC_2 = #x_536_2781
}

LET rnd(n) = VALOF 
{   
    LET val = (ranv!rani + ranv!ranj) & MASK
    ranv!rani := val
    rani := (rani + 1) MOD SIZE
    ranj := (ranj + 1) MOD SIZE
    RESULTIS val MOD n
}

AND initrnd(seed) = VALOF 
{   
    LET a, b = MAGIC_1 + seed, MAGIC_2
    FOR i = 0 TO (SIZE - 1) DO
    { 
        LET t = (a+b) & MASK
        a := b
        b := t
        ranv!i := t
    }
    rani, ranj := 0, 31
    RESULTIS TRUE
}

AND ranv = VEC (SIZE - 1)