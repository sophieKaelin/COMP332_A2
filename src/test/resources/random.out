Program(
    Vector(
        ManifestDecl(
            Vector(
                ManifestEntry(IdnDef("SIZE"), Some(IntExp(55))),
                ManifestEntry(IdnDef("MASK"), Some(IntExp(268435455))),
                ManifestEntry(IdnDef("MAGIC_1"), Some(IntExp(36984440))),
                ManifestEntry(IdnDef("MAGIC_2"), Some(IntExp(87435137))))),
        LetDecl(
            Vector(
                LetFnClause(
                    IdnDef("rnd"),
                    Vector(IdnDef("n")),
                    ValofExp(
                        Block(
                            Vector(
                                LetDecl(
                                    Vector(
                                        LetVarClause(
                                            Vector(IdnDef("val")),
                                            Vector(
                                                AndExp(
                                                    PlusExp(
                                                        BinaryPlingExp(
                                                            IdnExp(IdnUse("ranv")),
                                                            IdnExp(IdnUse("rani"))),
                                                        BinaryPlingExp(
                                                            IdnExp(IdnUse("ranv")),
                                                            IdnExp(IdnUse("ranj")))),
                                                    IdnExp(IdnUse("MASK")))))))),
                            Vector(
                                AssignStmt(
                                    Vector(
                                        BinaryPlingExp(
                                            IdnExp(IdnUse("ranv")),
                                            IdnExp(IdnUse("rani")))),
                                    Vector(IdnExp(IdnUse("val")))),
                                AssignStmt(
                                    Vector(IdnExp(IdnUse("rani"))),
                                    Vector(
                                        ModExp(
                                            PlusExp(IdnExp(IdnUse("rani")), IntExp(1)),
                                            IdnExp(IdnUse("SIZE"))))),
                                AssignStmt(
                                    Vector(IdnExp(IdnUse("ranj"))),
                                    Vector(
                                        ModExp(
                                            PlusExp(IdnExp(IdnUse("ranj")), IntExp(1)),
                                            IdnExp(IdnUse("SIZE"))))),
                                ResultIsStmt(ModExp(IdnExp(IdnUse("val")), IdnExp(IdnUse("n")))))))),
                LetFnClause(
                    IdnDef("initrnd"),
                    Vector(IdnDef("seed")),
                    ValofExp(
                        Block(
                            Vector(
                                LetDecl(
                                    Vector(
                                        LetVarClause(
                                            Vector(IdnDef("a"), IdnDef("b")),
                                            Vector(
                                                PlusExp(
                                                    IdnExp(IdnUse("MAGIC_1")),
                                                    IdnExp(IdnUse("seed"))),
                                                IdnExp(IdnUse("MAGIC_2"))))))),
                            Vector(
                                ForStmt(
                                    IdnDef("i"),
                                    IntExp(0),
                                    MinusExp(IdnExp(IdnUse("SIZE")), IntExp(1)),
                                    None,
                                    Block(
                                        Vector(
                                            LetDecl(
                                                Vector(
                                                    LetVarClause(
                                                        Vector(IdnDef("t")),
                                                        Vector(
                                                            AndExp(
                                                                PlusExp(
                                                                    IdnExp(IdnUse("a")),
                                                                    IdnExp(IdnUse("b"))),
                                                                IdnExp(IdnUse("MASK")))))))),
                                        Vector(
                                            AssignStmt(
                                                Vector(IdnExp(IdnUse("a"))),
                                                Vector(IdnExp(IdnUse("b")))),
                                            AssignStmt(
                                                Vector(IdnExp(IdnUse("b"))),
                                                Vector(IdnExp(IdnUse("t")))),
                                            AssignStmt(
                                                Vector(
                                                    BinaryPlingExp(
                                                        IdnExp(IdnUse("ranv")),
                                                        IdnExp(IdnUse("i")))),
                                                Vector(IdnExp(IdnUse("t"))))))),
                                AssignStmt(
                                    Vector(IdnExp(IdnUse("rani")), IdnExp(IdnUse("ranj"))),
                                    Vector(IntExp(0), IntExp(31))),
                                ResultIsStmt(TrueExp()))))),
                LetVecClause(IdnDef("ranv"), MinusExp(IdnExp(IdnUse("SIZE")), IntExp(1)))))))
